nomad/event_endpoint.go

```diff
@@ -0,0 +1,216 @@
+		if err != nil {
+			handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		}
+		return
+	}
+
+	// ACL check
+	// TODO(drew) ACL checks need to be per topic
+	// All Events        Management
+	// System Events     Management
+	// Node Events       NamespaceCapabilityReadEvents
+	// Job/Alloc Events  NamespaceCapabilityReadEvents
+	if aclObj, err := e.srv.ResolveToken(args.AuthToken); err != nil {
+		handleJsonResultError(err, nil, encoder)
+		return
+	} else if aclObj != nil && !aclObj.IsManagement() {
+		handleJsonResultError(structs.ErrPermissionDenied, helper.Int64ToPtr(403), encoder)
+		return
+	}
+
+	// authToken is passed to the subscribe request so the event stream
+	// can handle closing a subscription if the authToken expires.
+	// If ACLs are disabled, a random token is generated and it will
+	// never be closed due to expiry.
+	authToken := args.AuthToken
+	if authToken == "" {
+		authToken = uuid.Generate()
+	}
+	subReq := &stream.SubscribeRequest{
+		Token:     authToken,
+		Topics:    args.Topics,
+		Index:     uint64(args.Index),
+		Namespace: args.Namespace,
+	}
+	publisher, err := e.srv.State().EventBroker()
+	if err != nil {
+		handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// start subscription to publisher
+	subscription, err := publisher.Subscribe(subReq)
+	if err != nil {
+		handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+	defer subscription.Unsubscribe()
+
+	ndJsonCh := make(chan *structs.NDJson)
+	errCh := make(chan error)
+
+	jsonStream := stream.NewNDJsonStream(ndJsonCh, 30*time.Second)
+	jsonStream.Run(ctx)
+
+	// goroutine to detect remote side closing
+	go func() {
+		for {
+			if _, err := conn.Read(nil); err != nil {
+				if err == io.EOF || err == io.ErrClosedPipe {
+					// One end of the pipe was explicitly closed, exit cleanly
+					cancel()
+					return
+				}
+				select {
+				case errCh <- err:
+				case <-ctx.Done():
+					return
+				}
+			}
+		}
+	}()
+
+	go func() {
+		defer cancel()
+		for {
+			events, err := subscription.Next(ctx)
+			if err != nil {
+				select {
+				case errCh <- err:
+				case <-ctx.Done():
+				}
+				return
+			}
+
+			// Continue if there are no events
+			if len(events.Events) == 0 {
+				continue
+			}
+
+			if err := jsonStream.Send(events); err != nil {
+				select {
+				case errCh <- err:
+				case <-ctx.Done():
+				}
+				return
+			}
+		}
+	}()
+
+	var streamErr error
+OUTER:
+	for {
+		select {
+		case streamErr = <-errCh:
+			break OUTER
+		case <-ctx.Done():
+			break OUTER
+		case eventJSON, ok := <-ndJsonCh:
+			// check if ndjson may have been closed when an error occurred,
+			// check once more for an error.
+			if !ok {
+				select {
+				case streamErr = <-errCh:
+					// There was a pending error
+				default:
+				}
+				break OUTER
+			}
+
+			var resp structs.EventStreamWrapper
+			resp.Event = eventJSON
```

Comment:
I'm not sure if `NDJsonStream ` class and its goroutine management is worth its weight here.  Events are serialized and sent to this channel directly, and then we can have one more ticker select clause for heartbeats that encodes the same `structs.EventStreamWrapper` instance.

Reply:
I agree, and it may be a premature optimization but I think I'm going to use the (now JsonStream) NDJsonStream elsewhere for network sinks (that won't be run via RPC)